File: annotations-runtime/src/main/java/com/github/dynamicextensionsalfresco/workflow/activiti/WorkflowTaskRegistrar.java
package com.github.dynamicextensionsalfresco.workflow.activiti;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import org.activiti.engine.delegate.ExecutionListener;
import org.activiti.engine.delegate.JavaDelegate;
import org.activiti.engine.delegate.TaskListener;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;

/**
 * Detects workflow tasks from Spring beans implementing the {@link DelegateTask} or {@link TaskListener} interface.
 *
 * Still links components to the {@link WorkflowTaskRegistry} for backwards compatibility, but delegate expressions are
 * now the recommended way of referencing components from bpm files.
 *
 * @author Laurent Van der Linden
 */
public class WorkflowTaskRegistrar implements InitializingBean, ApplicationContextAware, DisposableBean {

    private final Logger logger = LoggerFactory.getLogger(WorkflowTaskRegistrar.class);

    private ApplicationContext applicationContext;

    @NotNull
    private final Map<String, Object> activitiBeanRegistry;
    @NotNull
    private final WorkflowTaskRegistry workflowTaskRegistry;

    public WorkflowTaskRegistrar(@NotNull Map<String, Object> activitiBeanRegistry,
            @NotNull WorkflowTaskRegistry workflowTaskRegistry) {
        if (activitiBeanRegistry == null) {
            throw new IllegalArgumentException("activitiBeanRegistry is null");
        }
        if (workflowTaskRegistry == null) {
            throw new IllegalArgumentException("workflowTaskRegistry is null");
        }

        this.activitiBeanRegistry = activitiBeanRegistry;
        this.workflowTaskRegistry = workflowTaskRegistry;
    }

    @Override
    public void afterPropertiesSet() {
        for (Class<?> type : workflowDelegateTypes()) {
            Map<String, ?> delegates = applicationContext.getBeansOfType(type);
            delegates.forEach((name, bean) -> {
                Object existing = activitiBeanRegistry.put(name, bean);
                if (existing != null) {
                    logger.warn("replaced existing {} with name {}", existing.getClass(), name);
                }
                logger.debug("registered Bundle component {} (type: {}) -> {}", name, type, bean);

                if (JavaDelegate.class.equals(type)) {
                    this.workflowTaskRegistry.registerDelegate(name, (JavaDelegate) bean);
                } else if (TaskListener.class.equals(type)) {
                    this.workflowTaskRegistry.registerTaskListener(name, (TaskListener) bean);
                } else if (ExecutionListener.class.equals(type)) {
                    this.workflowTaskRegistry.registerExecutionListener(name, (ExecutionListener) bean);
                }
            });
        }
    }

    @Override
    public void destroy() {
        for (Class<?> type : workflowDelegateTypes()) {
            Map<String, ?> delegates = applicationContext.getBeansOfType(type);
            delegates.forEach((name, bean) -> {
                activitiBeanRegistry.remove(name);
                logger.debug("unregistered Bundle component {} (type: {}) -> {}", name, type, bean);

                if (JavaDelegate.class.equals(type)) {
                    this.workflowTaskRegistry.unregisterDelegate(name);
                } else if (TaskListener.class.equals(type)) {
                    this.workflowTaskRegistry.unregisterTaskListener(name);
                } else if (ExecutionListener.class.equals(type)) {
                    this.workflowTaskRegistry.unregisterExecutionListener(name);
                }
            });
        }
    }

    @Override
    public void setApplicationContext(@NotNull ApplicationContext applicationContext) {
        if (applicationContext == null) {
            throw new IllegalArgumentException("applicationContext is null");
        }
        this.applicationContext = applicationContext;
    }

    @NotNull
    public final Map getActivitiBeanRegistry() {
        return this.activitiBeanRegistry;
    }

    @NotNull
    public final WorkflowTaskRegistry getWorkflowTaskRegistry() {
        return this.workflowTaskRegistry;
    }

    private List<Class<?>> workflowDelegateTypes() {
        return Arrays.asList(JavaDelegate.class, TaskListener.class, ExecutionListener.class);
    }
}

File: annotations-runtime/src/main/kotlin/com/github/dynamicextensionsalfresco/workflow/activiti/WorkflowTaskRegistrar.kt
package com.github.dynamicextensionsalfresco.workflow.activiti

import com.github.dynamicextensionsalfresco.debug
import com.github.dynamicextensionsalfresco.warn
import org.activiti.engine.delegate.ExecutionListener
import org.activiti.engine.delegate.JavaDelegate
import org.activiti.engine.delegate.TaskListener
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.DisposableBean
import org.springframework.beans.factory.InitializingBean
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware

/**
 * Detects workflow tasks from Spring beans implementing the [DelegateTask] or [TaskListener] interface.
 *
 * Still links components to the [WorkflowTaskRegistry] for backwards compatibility, but delegate expressions are now
 * the recommended way of referencing components from bpm files.

 * @author Laurent Van der Linden
 */
public open class WorkflowTaskRegistrar(val activitiBeanRegistry: MutableMap<String, Any>, val workflowTaskRegistry: WorkflowTaskRegistry)
        : InitializingBean, ApplicationContextAware, DisposableBean {
    private val logger = LoggerFactory.getLogger(WorkflowTaskRegistrar::class.java)

    private lateinit var applicationContext: ApplicationContext

    private fun workflowDelegateTypes() = listOf(
            JavaDelegate::class.java, TaskListener::class.java, ExecutionListener::class.java
    )

    override fun setApplicationContext(applicationContext: ApplicationContext) {
        this.applicationContext = applicationContext
    }

    override fun afterPropertiesSet() {
        for (type in workflowDelegateTypes()) {
            val delegates = applicationContext.getBeansOfType(type)
            for ((name,bean) in delegates) {
                val existing = activitiBeanRegistry.put(name, bean)
                if (existing != null) {
                    logger.warn { "replaced existing ${existing.javaClass} with name $name"}
                }
                logger.debug { "registered Bundle component $name (type: $type) -> $bean" }

                // legacy
                when (bean) {
                    is JavaDelegate -> workflowTaskRegistry.registerDelegate(name, bean)
                    is TaskListener -> workflowTaskRegistry.registerTaskListener(name, bean)
                    is ExecutionListener -> workflowTaskRegistry.registerExecutionListener(name, bean)
                }
            }
        }
    }

    override fun destroy() {
        for (type in workflowDelegateTypes()) {
            val delegates = applicationContext.getBeansOfType(type)
            for ((name, bean) in delegates) {
                activitiBeanRegistry.remove(name)
                logger.debug { "unregistered Bundle component $name (type: $type) -> $bean" }

                // legacy
                when (bean) {
                    is JavaDelegate -> workflowTaskRegistry.unregisterDelegate(name)
                    is TaskListener -> workflowTaskRegistry.unregisterTaskListener(name)
                    is ExecutionListener -> workflowTaskRegistry.unregisterExecutionListener(name)
                }
            }
        }
    }
}
